h2o-algos/src/main/java/hex/gram/Gram.java

```diff
@@ -667,6 +669,28 @@ public final void addRow(DataInfo.Row row, double w) {
     else
       addRowSparse(row, w);
   }
+
+  public final void addElementIJ(DataInfo.Row rowi, double wi, DataInfo.Row rowj, double wj, int ithRow, int jthRow,
+                                 int[] catOffsets) {
+    // nums
+    double elementij = 0.0;
+    double weights = wi*wj;
+    for(int i = 0; i < _denseN; ++i)  {
+      elementij += weights*rowi.numVals[i]*rowj.numVals[i]; // multiply numerical parts of columns
+    }
+
+    // cat X cat
+    if (catOffsets != null) {
```

Comment:
I don't understand this code. Are we doing dot product of the rows? If so the categorical part is a dot product of two spare binary vectors. Meaning which should just sum up the ids appearing in both? in that case, we need to do a variation of a sorted merge, i.e. need two indexes and only advance i if rowj.binIds[j] is bigger and vice versa. Why are we adding the cardinality? Or do I get it wrong ?

Also, if we're doing dot product of the two rows, I'd pu it into a Row.dot(Row) method.